lib/Service/Social/CompositeSocialProvider.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * @copyright 2020 Matthias Heinisch <nextcloud@matthiasheinisch.de>
+ *
+ * @author Matthias Heinisch <nextcloud@matthiasheinisch.de>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Contacts\Service\Social;
+
+/**
+ * Composition of all social providers for easier usage
+ */
+class CompositeSocialProvider {
+
+	/** @var ISocialProvider[] */
+	private $providers;
+
+	public function __construct(InstagramProvider $instagramProvider,
+					MastodonProvider $mastodonProvider,
+					FacebookProvider $facebookProvider,
+					TwitterProvider $twitterProvider,
+					TumblrProvider $tumblrProvider) {
+
+		// This determines the priority of known providers
+		$this->providers = [
+			'instagram' 	=> $instagramProvider,
+			'mastodon' 	=> $mastodonProvider,
+			'twitter' 	=> $twitterProvider,
+			'facebook'	=> $facebookProvider,
+			'tumblr'	=> $tumblrProvider,
+		];
+	}
+
+	/**
+	 * returns an array of supported social providers
+	 *
+	 * @returns {array} array of the supported social networks
+	 */
+	public function getSupportedNetworks() : array {
+		return array_keys($this->providers);
+	}
+
+
+	/**
+	 * generate download url for a social entry
+	 *
+	 * @param {array} socialEntries all social data from the contact
+	 * @param {String} network the choice which network to use (fallback: take first available)
+	 *
+	 * @returns {String} the url to the requested information or null in case of errors
+	 */
+	public function getSocialConnector(array $socialEntries, string $network) : ?string {
+
+		$connector = null;
+		$selection = $this->providers;
+		// check if dedicated network selected
+		if (isset($this->providers[$network])) {
+			$selection = array($network => $this->providers[$network]);
+		}
+
+		// check selected providers in order
+		foreach($selection as $type => $socialProvider) {
+
+			// search for this network in user's profile
+			foreach ($socialEntries as $socialEntry) {
+
+				if (strtolower($type) === strtolower($socialEntry['type'])) {
+					$profileId = $socialProvider->cleanupId($socialEntry['value']);
+					if (!is_null($profileId)) {
+						$connector = $socialProvider->getImageUrl($profileId);
```

Comment:
simply `return` here and leave out all the `break`s. No need to finish the iteration of the loop when we don't want to do anything else.

Reply:
The idea was that all supported networks are searched for in a defined order. If that social network has an entry, the connector is calculated - but it can happen that there is a problem with the id or the connectivity to that network, so the connector becomes null. In that case the loop shall continue with the next supported network. But if the connector is not null, a break quits the loop. Normally there should not be any unnecessary iterations present.

Reply:
I see. In that case you should indeed continue the iteration. I naively assumed that the `getImageUrl` will always return a result. If that also allows `null` then the current code makes sense.

Reply:
Yes, this can happen in fact. For example the Instagram API returns a new imageUrl every time, so its not a static recipe.